### PR TITLE
Adding Pinger interface and Removing mandatory nature of Ping fuction

### PR DESCRIPTION
**Pull request by DeepanshuA**
# Description

This change introduces Pinger interface and also reduces mandatory Ping functionality.
Ping works on the basis, that wen called, it returns either an Error or nil.
With forced implementation, any component returning without checking actual heartbeat and simply returning nil, causes wrong sense of correctness of component. So, ideally Ping should be implemented only when correctly implemented.

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: https://github.com/dapr/dapr/issues/2167

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by hueifeng** [comment]
Please fix DCO.

File: state/store.go [hunk 1]
```diff
@@ -21,7 +27,15 @@ type Store interface {
...
 	Set(req *SetRequest) error
-	Ping() error
+}
+
+func Ping(store Store) error {
+	// checks if this store has the ping option then executes
+	if storeWithPing, ok := store.(health.Pinger); ok {
+		return storeWithPing.Ping()
+	} else {
+		return fmt.Errorf("Ping is not implemented by this store")
```
Comment (berndverst) [inline state/store.go:37]:
```suggestion
		return fmt.Errorf("Ping is not implemented by this state store")
```

File: tests/conformance/state/state.go [hunk 2]
```diff
@@ -227,8 +227,15 @@ func ConformanceTests(t *testing.T, props map[string]string, statestore state.St
...
 
 	t.Run("ping", func(t *testing.T) {
-		err := statestore.Ping()
-		assert.Nil(t, err)
+		err := state.Ping(statestore)
+		// TODO: Ideally, all stable components should implenment ping function,
+		// so will only assert assert.Nil(t, err) finally, i.e. when current implementation
+		// implements ping in existing stable components
+		if err != nil {
+			assert.EqualError(t, err, "Ping is not implemented by this store")
```
Comment (berndverst) [inline tests/conformance/state/state.go:235]:
```suggestion
			assert.EqualError(t, err, "Ping is not implemented by this state store")
```

File: bindings/input_binding.go [hunk 3]
```diff
@@ -25,3 +30,12 @@ type InputBinding interface {
 
 // Handler is the handler used to invoke the app handler.
 type Handler func(context.Context, *ReadResponse) ([]byte, error)
+
+func PingInpBinding(inputBinding InputBinding) error {
+	// checks if this input binding has the ping option then executes
+	if inputBindingWithPing, ok := inputBinding.(health.Pinger); ok {
+		return inputBindingWithPing.Ping()
+	} else {
+		return fmt.Errorf("Ping is not implemented by this Input Binding")
```
Comment (berndverst) [inline bindings/input_binding.go:39]:
```suggestion
		return fmt.Errorf("Ping is not implemented by this input binding")
```

this should be lowercase for consistency and it currently breaks tests

File: bindings/output_binding.go [hunk 4]
```diff
@@ -13,11 +13,25 @@ limitations under the License.
...
 	Invoke(ctx context.Context, req *InvokeRequest) (*InvokeResponse, error)
 	Operations() []OperationKind
 }
+
+func PingOutBinding(outputBinding OutputBinding) error {
+	// checks if this output binding has the ping option then executes
+	if outputBindingWithPing, ok := outputBinding.(health.Pinger); ok {
+		return outputBindingWithPing.Ping()
+	} else {
+		return fmt.Errorf("Ping is not implemented by this Output Binding")
```
Comment (berndverst) [inline bindings/output_binding.go:35]:
```suggestion
		return fmt.Errorf("Ping is not implemented by this output binding")
```

File: pubsub/pubsub.go [hunk 5]
```diff
@@ -26,3 +31,12 @@ type PubSub interface {
 
 // Handler is the handler used to invoke the app handler.
 type Handler func(ctx context.Context, msg *NewMessage) error
+
+func Ping(pubsub PubSub) error {
+	// checks if this pubsub has the ping option then executes
+	if pubsubWithPing, ok := pubsub.(health.Pinger); ok {
+		return pubsubWithPing.Ping()
+	} else {
+		return fmt.Errorf("Ping is not implemented by this PubSub")
```
Comment (berndverst) [inline pubsub/pubsub.go:40]:
```suggestion
		return fmt.Errorf("Ping is not implemented by this pubsub")
```

**Review by berndverst — CHANGES_REQUESTED** [review]
Please also fix DCO

File: tests/conformance/state/state.go (hunk 2 above)
Comment (DeepanshuA) [reply tests/conformance/state/state.go]: Done

File: bindings/input_binding.go (hunk 3 above)
Comment (DeepanshuA) [reply bindings/input_binding.go]: Done. To understand that what tests are failing and how these rules are setup, I tried to search, are these the Conformance tests that are failing? I could not find this in UTs.

File: bindings/output_binding.go (hunk 4 above)
Comment (DeepanshuA) [reply bindings/output_binding.go]: Done

File: pubsub/pubsub.go (hunk 5 above)
Comment (DeepanshuA) [reply pubsub/pubsub.go]: Done

**Review by berndverst — APPROVED** [review]

**Pull request merged**